java/src/org/openqa/selenium/grid/node/config/SessionCapabilitiesMutator.java

```diff
@@ -56,33 +57,137 @@ public Capabilities apply(Capabilities capabilities) {
+
+      @SuppressWarnings("unchecked")
+      Map<String, Object>
+        stereotypeOptions =
+        (Map<String, Object>) slotStereotype.asMap().get(options);
+
+      @SuppressWarnings("unchecked")
+      Map<String, Object> capsOptions = (Map<String, Object>) capabilities.asMap().get(options);
+
+      // Merge top level capabilities, excluding browser specific options.
+      // This will overwrite the browser options too, but it does not matter since we tackle it separately just after this.
+      Map<String, Object> toReturn = new HashMap<>(slotStereotype.merge(capabilities).asMap());
+
+      // Merge browser specific stereotype and capabilities options
+      switch (browserName) {
+        case "chrome":
+        case "microsoftedge":
+          toReturn.put(options, mergeChromiumOptions(stereotypeOptions, capsOptions));
+          break;
+        case "firefox":
+          toReturn.put(options, mergeFirefoxOptions(stereotypeOptions, capsOptions));
+          break;
+        default:
+          break;
+      }
+
+      return new ImmutableCapabilities(toReturn);
     }
 
-    @SuppressWarnings("unchecked")
-    Map<String, Object> stereotypeOptions = (Map<String, Object>) slotStereotype.asMap().get(options);
+    return slotStereotype.merge(capabilities);
+  }
+
+  private Map<String, Object> mergeChromiumOptions(Map<String, Object> stereotypeOptions,
+                                                   Map<String, Object> capsOptions) {
+    Map<String, Object> toReturn = new HashMap<>(stereotypeOptions);
 
-    Map<String, Object> toReturn = new HashMap<>(capabilities.asMap());
+    for (String name : capsOptions.keySet()) {
+      if (name.equals("args")) {
+        List<String> arguments =
+          (List<String>) (capsOptions.getOrDefault(("args"), new ArrayList<>()));
 
-    if (!toReturn.containsKey(options)) {
-      toReturn.put(options, stereotypeOptions);
-      return new ImmutableCapabilities(toReturn);
+        List<String> stereotypeArguments =
+          (List<String>) (stereotypeOptions.getOrDefault(("args"), new ArrayList<>()));
+
+        arguments.forEach(arg -> {
+          if (!stereotypeArguments.contains(arg)) {
+            stereotypeArguments.add(arg);
+          }
+        });
+        toReturn.put("args", stereotypeArguments);
+      }
+
+      if (name.equals("extensions")) {
+        List<String> extensionList = (List<String>) (capsOptions.get(("extensions")));
+
+        List<String> stereotypeExtensions =
+          (List<String>) (stereotypeOptions.getOrDefault(("extensions"), new ArrayList<>()));
+
+        extensionList.forEach(extension -> {
+          if (!stereotypeExtensions.contains(extension)) {
+            stereotypeExtensions.add(extension);
+          }
+        });
+
+        toReturn.put("extensions", stereotypeExtensions);
+      }
+
+      if (name.equals("binary")) {
+        Object binary = capsOptions.get("binary");
+        toReturn.put("binary", binary);
+      }
```

Comment:
Somehow this comment was lost...

If I understand the code correctly, this would overwrite the `binary` location in the stereotype (if it exists), which is probably not ideal because the person who setup the Grid knows the browser paths.

Reply:
Ahh okay, I understand now. Thank you for helping clarify this. I will fix it.